milli/src/search/criteria/mod.rs

```diff
@@ -318,21 +319,41 @@ pub fn resolve_query_tree<'t>(
             }
             Phrase(words) => {
                 let mut candidates = RoaringBitmap::new();
-                let mut first_loop = true;
-                for slice in words.windows(2) {
-                    let (left, right) = (&slice[0], &slice[1]);
-                    match ctx.word_pair_proximity_docids(left, right, 1)? {
-                        Some(pair_docids) => {
-                            if pair_docids.is_empty() {
-                                return Ok(RoaringBitmap::new());
-                            } else if first_loop {
-                                candidates = pair_docids;
-                                first_loop = false;
-                            } else {
-                                candidates &= pair_docids;
-                            }
+                let mut first_iter = true;
+                let winsize = words.len().min(7);
+
+                for win in words.windows(winsize) {
+                    // Get all the word pairs and their compute their relative distance
+                    let dists = win
+                        .iter()
+                        .enumerate()
+                        .cartesian_product(win.iter().enumerate())
+                        .filter(|(x, y)| y > x)
+                        .map(|((pos1, s1), (pos2, s2))| (s1, s2, pos2 - pos1));
+
+                    let mut bitmaps = Vec::with_capacity(winsize.pow(2));
+
+                    for (s1, s2, d) in dists {
+                        match ctx.word_pair_proximity_docids(s1, s2, d as u8)? {
+                            Some(m) => bitmaps.push(m),
+                            None => return Ok(RoaringBitmap::new()),
+                        }
+                    }
```

Comment:
What do you think about something like that ? 🤔 
```suggestion
                for mut win in words.windows(winsize) {
                    let mut bitmaps = Vec::with_capacity(winsize.pow(2));

                    while let Some((head, tail)) = win.split_first() {
                        for (dist, w) in tail.iter().enumerate() {
                            match ctx.word_pair_proximity_docids(s1, s2, dist as u8 + 1)? {
                                Some(m) => bitmaps.push(m),
                                None => return Ok(RoaringBitmap::new()),
                            }
                        }

                        win = tail;
                    }
```

Reply:
unless you have strong reasons, I would favour the declarative approach

Reply:
I find it shorter, this avoids doing 3 loops and a cartesian product with a filter is not so clear.
But I understand that the first cartesian product has "no cost" and the comment explains its purpose.

Reply:
Indeed, I find the Many's solution easier to understand and expressing the behavior better than having to understand that the Cartesian product is not used as is and values are filtered out afterwards.

Reply:
wdyt about that: 
```rust
                    let mut bitmaps = Vec::with_capacity(winsize.pow(2));
                    for (offset, s1) in win.iter().enumerate() {
                        for (dist, s2) in win.iter().skip(offset + 1).enumerate() {
                            match ctx.word_pair_proximity_docids(s1, s2, dist as u8 + 1)? {
                                Some(m) => bitmaps.push(m),
                                None => return Ok(RoaringBitmap::new()),
                            }
                        }
                    }
```

we avoid the mutable window this way, which is a bit awkward.

Reply:
@MarinPostma, yes, this version is the clearest.
Meanwhile, I have a rust question, does the implementation `win.iter().skip(offset + 1).enumerate()` is the same of `win[offset + 1..].into_iter().enumerate()` in term of time?

Reply:
yes, but the former can't panic